jlm/llvm/opt/alias-analyses/PointerObjectSet.cpp

```diff
@@ -0,0 +1,303 @@
+    modified |= GetPointerObject(pointee).MarkAsEscaped();
+  return modified;
+}
+
+
+// P(superset) is a superset of P(subset)
+bool
+SupersetConstraint::Apply(PointerObjectSet& set)
+{
+  return set.MakePointsToSetSuperset(Superset_, Subset_);
+}
+
+// for all x in P(pointer1), make P(x) a superset of P(pointer2)
+bool
+AllPointeesPointToSupersetConstraint::Apply(PointerObjectSet& set)
+{
+  bool modified = false;
+
+  for (PointerObject::Index x : set.GetPointsToSet(Pointer1_)) {
+    modified |= set.MakePointsToSetSuperset(x, Pointer2_);
+  }
+
+  // If external in P(Pointer1_), P(external) should become a superset of P(Pointer2)
+  // In practice, this means everything in P(Pointer2) escapes
+  if (set.GetPointerObject(Pointer1_).PointsToExternal())
+    modified |= set.MarkAllPointeesAsEscaped(Pointer2_);
+
+  return modified;
+}
+
+// Make P(loaded) a superset of P(x) for all x in P(pointer)
+bool
+SupersetOfAllPointeesConstraint::Apply(PointerObjectSet& set)
+{
+  bool modified = false;
+
+  for (PointerObject::Index x : set.GetPointsToSet(Pointer_)) {
+    modified |= set.MakePointsToSetSuperset(Loaded_, x);
+  }
+
+  // Handling pointing to external is done by MakePointsToSetSuperset,
+  // Propagating escaped status is handled by different constraints
+
+  return modified;
+}
+
+
+void
+PointerObjectConstraintSet::AddPointerPointeeConstraint(PointerObject::Index pointer, PointerObject::Index pointee)
+{
+  // All set constraints are additive, so simple constraints like this can be directly applied and forgotten.
+  Set_.AddToPointsToSet(pointer, pointee);
+}
+
+void
+PointerObjectConstraintSet::AddPointsToExternalConstraint(PointerObject::Index pointer)
+{
+  // Flags are never removed, so adding the flag now ensures it will be included in the final solution
+  Set_.GetPointerObject(pointer).MarkAsPointsToExternal();
+}
+
+void
+PointerObjectConstraintSet::AddRegisterContentEscapedConstraint(PointerObject::Index registerIndex)
+{
+  // Registers themselves can't really escape, since they don't have an address
+  // We can however mark it as escaped, and let escape flag propagation ensure everything it ever points to is marked.
+  auto& registerPointerObject = Set_.GetPointerObject(registerIndex);
+  JLM_ASSERT(registerPointerObject.GetKind() == PointerObjectKind::Register);
+  registerPointerObject.MarkAsEscaped();
+}
+
+void
+PointerObjectConstraintSet::AddConstraint(ConstraintVariant c)
+{
+  Constraints_.push_back(c);
+}
+
+void
+PointerObjectConstraintSet::Solve() {
+  // Keep applying constraints until no sets are modified
+  bool modified = true;
+
+  while (modified) {
+    modified = false;
+
+    for (auto& constraint : Constraints_)
+      std::visit([&](auto constraint) {
+        modified |= constraint.Apply(Set_);
+      }, constraint);
+
+    modified |= PropagateEscapedFlag();
+  }
+}
+
+bool
+PointerObjectConstraintSet::PropagateEscapedFlag()
+{
+  bool modified = false;
```

Comment:
Could we move this line of code between the while and for loop. It is not used in the first part of the code, i.e., the while loop.